cmd/fts/cmd/tasks.go

```diff
@@ -0,0 +1,232 @@
+		close(res)
+	}()
+
+	return res
+}
+
+func ProcessTask(task Task, config RunConfig, ch chan bool, res chan<- TaskResult, wg *sync.WaitGroup, sc *StagingConfig) {
+	defer wg.Done()
+
+	// Fill in the channel buffer slot.
+	ch <- true
+
+	taskResult := TaskResult{
+		task:  task,
+		stage: "waiting",
+	}
+
+	sendErr := func(err error) {
+		taskResult.err = err
+		res <- taskResult
+	}
+
+	for {
+		// check available space
+		ready, err := sc.Ready(task.bytes)
+		if err != nil {
+			sendErr(err)
+			break
+		}
+		// wait for space
+		if !ready {
+			timeout := 1000 * time.Millisecond
+			time.Sleep(timeout)
+			continue
+		}
+
+		Message("Starting storage of %s.", task.path)
+		taskResult.stage = "started"
+		// TODO: allow user to set target at input
+		pow, err := client.NewClient(config.serverAddress)
+		if err != nil {
+			sc.Done(task.bytes)
+			sendErr(err)
+			break
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Hour*8)
+		defer cancel()
+		ctx = context.WithValue(ctx, client.AuthKey, config.token)
+
+		if config.dryRun {
+			taskResult.cid = "dry-run"
+			taskResult.stage = "dry-run"
+			sc.Done(task.bytes)
+			res <- taskResult
+			break
+		}
+
+		if task.isDir {
+			taskResult.cid, err = pow.Data.StageFolder(ctx, config.ipfsrevproxy, task.path)
+			if err != nil {
+				sc.Done(task.bytes)
+				sendErr(err)
+				break
+			}
+		} else {
+			f, err := os.Open(task.path)
+			if err != nil {
+				sc.Done(task.bytes)
+				sendErr(err)
+				break
+			}
+			stageRes, err := pow.Data.Stage(ctx, f)
+			if err != nil {
+				sc.Done(task.bytes)
+				sendErr(err)
+				break
+			}
+			taskResult.cid = stageRes.Cid
+			if err := f.Close(); err != nil {
+				sc.Done(task.bytes)
+				sendErr(err)
+				break
+			}
+		}
+
+		taskResult.stage = "stagged"
+
+		if config.dryRun {
+			taskResult.stage = "dry-run"
+			sc.Done(task.bytes)
+			break
+		}
+
+		applyRes, err := pow.StorageConfig.Apply(ctx, taskResult.cid)
+		if err != nil {
+			sc.Done(task.bytes)
+			sendErr(err)
+			break
+		}
+		taskResult.jobId = applyRes.JobId
+
+		taskResult.stage = "queued"
+
+		Message("Waiting to complete deals for %s.", task.path)
+
+		storage, err := watchStorage(pow, config.token, taskResult.jobId)
+		if err != nil {
+			sc.Done(task.bytes)
+			sendErr(err)
+			break
+		}
+		taskResult.state = storage
+
+		Message("Completed storage of %s.", task.path)
+
+		var opts []client.DealRecordsOption
+		opts = append(opts, client.WithDataCids(taskResult.cid))
+		opts = append(opts, client.WithIncludeFinal(true))
+		opts = append(opts, client.WithIncludePending(false))
+		complete, err := pow.Deals.StorageDealRecords(ctx, opts...)
+		if err != nil {
+			sc.Done(task.bytes)
+			sendErr(err)
+			break
+		}
+		taskResult.records = complete.Records
+		sc.Done(task.bytes)
+		res <- taskResult
+		break
+	}
+	// Free the channel buffer slot.
+	<-ch
+}
+
+func watchStorage(pow *client.Client, token string, jobId string) (*client.WatchStorageJobsEvent, error) {
```

Comment:
We should probably change this implementation. The main problem with using stream APIs is that they have a timeout if no new data is streamed. This is a safety measure in gRPC to avoid long-lived idle connections that can lead to port exhaustion. In the case of Powergate it's highly probable that the connection will be closed and you reach the final return with error, since it can pass multiple hours without new messages.

[In Hub](https://github.com/textileio/textile/blob/master/buckets/archive/tracker.go) we do polling, with shorter intervals for initial stages, and longer ones for "Sealing". 
So basically a for loop with a sleep in this case. In Hub we do something a bit more sophisticated to allow killing the Hub process and let resuming all needed things.

Reply:
The "watch log" API maybe was too biased for localnet usage since it's quite fast and has a great UX. But for real usage, it has limited application and might be even misleading as happened here. So something to think about...

Reply:
cleaned that up and created a watching channel that will poll for updates as long as deals are still out there